packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -0,0 +1,304 @@
+    let index = 0;
+    for (let [key, node] of this.keyMap) {
+      if (last) {
+        last.nextKey = key;
+        node.prevKey = last.key;
+      } else {
+        this.firstKey = key;
+      }
+
+      if (node.type === 'item') {
+        node.index = index++;
+      }
+
+      last = node;
+      // Set nextKey as undefined since this might be the last node
+      // If it isn't the last node, last.nextKey will properly set at start of new loop 
+      last.nextKey = undefined;
+    }
+
+    this.lastKey = last ? last.key : ''; // what to do in empty collection??
+  }
+
+  *[Symbol.iterator]() {
+    yield* filter(this.iterable, this.filterFn);
+  }
+
+  get size() {
+    return this.keyMap.size;
+  }
+
+  getKeys() {
+    return this.keyMap.keys();
+  }
+
+  getKeyBefore(key: Key) {
+    let node = this.keyMap.get(key);
+    return node ? node.prevKey : null;
+  }
+
+  getKeyAfter(key: Key) {
+    let node = this.keyMap.get(key);
+    return node ? node.nextKey : null;
+  }
+
+  getFirstKey() {
+    return this.firstKey;
+  }
+
+  getLastKey() {
+    return this.lastKey;
+  }
+
+  getItem(key: Key) {
+    return this.keyMap.get(key);
+  }
+}
+
+let whitespace = /\s/;
+
+export function useComboBoxState<T extends object>(props: ComboBoxProps<T>): ComboBoxState<T> {
+  let {
+    onFilter,
+    selectedKey,
+    defaultSelectedKey,
+    inputValue,
+    defaultInputValue,
+    onInputChange,
+    isFocused,
+    isOpen,
+    defaultOpen
+  } = props;
+  
+  let itemsControlled = !!onFilter;
+  let collator = useCollator({sensitivity: 'base'});
+  /*
+  let menuControlled = props.isOpen !== undefined;
+  let valueControlled = inputValue !== undefined;
+  let selectedControlled = !!selectedKey;
+   */
+
+  // Create a separate menu open state tracker so onOpenChange doesn't fire with open and close in quick succession
+  // in cases where there aren't items to show
+  // Note that this means onOpenChange won't fire for controlled open states
+  let [menuIsOpen, setMenuIsOpen] = useControlledState(isOpen, defaultOpen || false, () => {});
+
+  let selectState = useSelectState(props);
+
+  let selectedKeyItem = selectedKey ? selectState.collection.getItem(selectedKey) : undefined;
+  let selectedKeyText = selectedKeyItem ? selectedKeyItem.textValue || selectedKeyItem.rendered as string : undefined;
+  // Maybe don't need to do for defaultSelectedKey? Need to do for selectedKey so that the textfield is properly controlled reflects selectedKey text
+  let defaultSelectedKeyItem = defaultSelectedKey ? selectState.collection.getItem(defaultSelectedKey) : undefined;
+  let defaultSelectedKeyText = defaultSelectedKeyItem ? defaultSelectedKeyItem.textValue || defaultSelectedKeyItem.rendered as string : undefined;
+  // Double check if selectedKey should make textfield value controlled
+  let [value, setValue] = useControlledState(toString(inputValue) || selectedKeyText, toString(defaultInputValue) || defaultSelectedKeyText || '', onInputChange);
```

Comment:
I think I would not try to combine the text from `inputValue` and `selectedKeyText` in state. It's going to make things complicated. Instead, just store `inputValue` in state, and the `selectedKey` in state, and then set the value you return from `useComboBoxState` below as `inputValue || selectedKeyText`. You should also throw an error if they both exist and don't match.